R/write_pxl_file.R

```diff
@@ -0,0 +1,499 @@
+#'  - Sample meta data : The sample meta data is extracted from the original .pxl file(s),
+#'  then merged and exported to a .json file.
+#'
+#' The structure of the .pxl file is detailed below:
+#'
+#' |-- adata.h5ad\cr
+#' |-- polarization.parquet\cr
+#' |-- colocalization.parquet\cr
+#' |-- metadata.json\cr
+#' |-- edgelist.parquet
+#'
+#' The merged files are converted into a zip archive and saved to the target .pxl file.
+#'
+#' @param object A \code{Seurat} object with a \code{CellGraphAssay5}
+#' assay object created with pixelatorR.
+#' @param file A character string specifying the path to the .pxl
+#' file to be created.
+#' @param assay A character string specifying the name of the \code{CellGraphAssay5}.
+#' If set to \code{NULL} the default assay will be used.
+#' @param overwrite A logical value specifying whether to overwrite the \code{file}
+#' if it already exists.
+#'
+#' @return Nothing. The function writes the .pxl file to the specified location.
+#'
+#' @examples
+#' # Use Assay5 as the default assay version
+#' options(Seurat.object.assay.version = "v5")
+#'
+#' # Create Seurat object
+#' pxl_file <- system.file("extdata/five_cells",
+#'                         "five_cells.pxl",
+#'                         package = "pixelatorR")
+#' se <- ReadMPX_Seurat(pxl_file)
+#'
+#' se_merged <- merge(se, list(se, se, se))
+#' pxl_file <- fs::path_temp("small.pxl")
+#'
+#' # Export data to a new .pxl file
+#' WriteMPX_pxl_file(se_merged, pxl_file)
+#'
+#' # Read the new .pxl file
+#' se_merged <- ReadMPX_Seurat(pxl_file)
+#'
+#' # Reset global option
+#' options(Seurat.object.assay.version = "v3")
+#'
+#' @export
+#'
+WriteMPX_pxl_file <- function (
+  object,
+  file,
+  assay = NULL,
+  overwrite = FALSE
+) {
+
+  stopifnot(
+    "'object' must be a Seurat object" =
+      inherits(object, what = "Seurat"),
+    "'file' must be a valid path ending with .pxl" =
+      is.character(file) && (fs::path_ext(file) == "pxl"),
+    "'overwrite' must be either TRUE or FALSE" =
+      is.logical(overwrite)
+  )
+
+  # Check if file exists
+  if (fs::file_exists(file) & !overwrite) {
+    abort(glue("{col_br_blue(file)} already exists. Please select a different ",
+               "file name or set overwrite = TRUE if you are certain that the ",
+               "existing file should be replaced."))
+  }
+
+  assay <- assay %||% DefaultAssay(object)
+  cg_assay <- object[[assay]]
+  stopifnot(
+    "'assay' must be a 'CellGraphAssay' or a 'CellGraphAssay5' object" =
+      is(cg_assay, "MPXAssay")
+  )
+
+  # fetch and validate fs_map
+  fs_map <- cg_assay@fs_map
+  .validate_fs_map(fs_map)
+
+  # Delete the file only if it's not in use
+  # by the current object
+  if (fs::file_exists(file)) {
+    if (file %in% fs_map$pxl_file) {
+      abort(glue(
+        "The selected file name '{col_br_blue(file)}' is currently in use ",
+        "by the input object. You need to select a different file name."
+      ))
+    }
+    fs::file_delete(file)
```

Comment:
A general comment; why is `fs` used instead of base functions?

Reply:
We could switch to base R, but I prefer fs because it is IMO more rigorous. For example:

"Compared to [file.remove](http://127.0.0.1:10927/help/library/fs/help/file.remove) they always fail if they cannot delete the object rather than changing return value or signalling a warning."

We could of course handle warnings manually, but I think this is a nice feature of fs that their function always errors if the command fails.

Reply:
Ok, that sounds like a good motivation to me!

Reply:
Ok then I'll keep it as is!